File: src/tools/rclone.ts
import { type ToolDefinition, type ToolResult } from "../core/types.js";

export function getTools(): ToolDefinition[] {
  const make = (n: string, d: string, s: Record<string, unknown>, fn: (p: any, ctx: any) => Promise<ToolResult>): ToolDefinition => ({
    name: n, description: d, inputSchema: s as any, execute: fn,
  });
  return [
    make("rclone_list_remotes", "Get configured remotes. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { rclone { listRemotes { name type } } }") })),
    make("rclone_config_form_schema", "Get remote config schema. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { rclone { configFormSchema } }") })),
    make("rclone_create_remote", "Create a remote.", { type:"object", properties:{ name:{type:"string"}, config:{type:"object"} }, required:["name","config"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation CreateRcloneRemote($name: String!, $config: JSON!) { rclone { createRemote(name: $name, config: $config) { name type } } }", p) })),
    make("rclone_delete_remote", "Delete a remote. Destructive.", { type:"object", properties:{ name:{type:"string"} }, required:["name"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation DeleteRcloneRemote($name: String!) { rclone { deleteRemote(name: $name) { status } } }", p) })),
  ];
}

File: go/internal/tools/rclone.go
package rclone

import (
	"context"
	"fmt"

	"github.com/hektyc/unraid-mcp-server/go/internal/config"
	"github.com/hektyc/unraid-mcp-server/go/internal/guards"
	"github.com/hektyc/unraid-mcp-server/go/internal/mcp"
)

func RegisterTools(s *mcp.Server, cfg *config.Config) {
	s.RegisterTool(&mcp.ToolDef{
		Name:        "rclone_list_remotes",
		Description: "Get configured remotes. Read-only.",
		Query:       `query { rclone { listRemotes { name type } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "rclone_create_remote",
		Description: "Create a remote.",
		Query:       `mutation CreateRcloneRemote($name: String!, $config: JSON!) { rclone { createRemote(name: $name, config: $config) { name type } } }`,
		Params: map[string]string{
			"name": "string", // required=true
			"config": "object", // required=true
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "rclone_delete_remote",
		Description: "Delete a remote.",
		Query:       `mutation DeleteRcloneRemote($name: String!) { rclone { deleteRemote(name: $name) { status } } }`,
		Params: map[string]string{
			"name": "string", // required=true
		},
	})
}